File: frontend/src/api/serviceGraph.ts
export const NODE_TYPES = ['service', 'topic', 'queue', 'database', 'external'] as const
export type NodeType = typeof NODE_TYPES[number]

export const EDGE_TYPES = ['rest', 'grpc', 'publish', 'subscribe', 'reads', 'writes', 'calls'] as const
export type EdgeType = typeof EDGE_TYPES[number]

export const NODE_TYPE_LABELS: Record<NodeType, string> = {
  service:    'Service',
  topic:      'Topic',
  queue:      'Queue',
  database:   'Database',
  external:   'External',
}

export const EDGE_TYPE_LABELS: Record<EdgeType, string> = {
  rest:       'REST',
  grpc:       'gRPC',
  publish:    'Publish',
  subscribe:  'Subscribe',
  reads:      'Reads',
  writes:     'Writes',
  calls:      'Calls',
}

export interface GraphNode {
  id: string
  node_type: NodeType
  name: string
  description: string
  tech: string
  position_x: number
  position_y: number
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  edge_type: EdgeType
  label: string
}

export interface ServiceGraph {
  nodes: GraphNode[]
  edges: GraphEdge[]
  updated_at: string
}

export async function fetchGraph(): Promise<ServiceGraph> {
  const res = await fetch('/api/v1/service-graph/')
  if (!res.ok) throw new Error(`Failed to load service graph: ${res.status}`)
  return res.json()
}

export async function saveGraph(nodes: GraphNode[], edges: GraphEdge[]): Promise<ServiceGraph> {
  const res = await fetch('/api/v1/service-graph/', {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ nodes, edges }),
  })
  if (!res.ok) throw new Error(`Failed to save service graph: ${res.status}`)
  return res.json()
}

File: backend/src/models/service_graph.py
"""Data models for the service graph (topology map)."""
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum

from pydantic import BaseModel, Field


class NodeType(str, Enum):
    service = "service"
    topic = "topic"
    queue = "queue"
    database = "database"
    external = "external"


class EdgeType(str, Enum):
    rest = "rest"
    grpc = "grpc"
    publish = "publish"
    subscribe = "subscribe"
    reads = "reads"
    writes = "writes"
    calls = "calls"


class GraphNode(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    node_type: NodeType = NodeType.service
    name: str
    description: str = ""
    tech: str = ""          # e.g. "Python/FastAPI", "Kafka", "PostgreSQL"
    position_x: float = 0.0
    position_y: float = 0.0


class GraphEdge(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    source: str             # node id
    target: str             # node id
    edge_type: EdgeType = EdgeType.rest
    label: str = ""


class ServiceGraph(BaseModel):
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )


class ServiceGraphUpdate(BaseModel):
    """Full replacement payload from the frontend editor."""
    nodes: list[GraphNode]
    edges: list[GraphEdge]
